File: algorithms/src/test/java/com/javaground/algorithms/trie/indexpairs/IndexPairsTest.java
package com.javaground.algorithms.trie.indexpairs;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.*;

class IndexPairsTest {

    public static Stream<Arguments> testCases() {
        return Stream.of(
                Arguments.of(
                        "thestoryofeducativeandme",
                        new String[]{"story", "feduc", "educative"},
                        new int[][]{{3, 7}, {9, 13}, {10, 18}}
                ),
                Arguments.of(
                        "xyxyx",
                        new String[]{"xyx", "xy"},
                        new int[][]{{0, 1}, {0, 2}, {2, 3}, {2, 4}}
                ),
                Arguments.of(
                        "howareyou",
                        new String[]{"how", "are", "you"},
                        new int[][]{{0, 2}, {3, 5}, {6, 8}}
                ),
                Arguments.of(
                        "weather",
                        new String[]{"weather"},
                        new int[][]{{0, 6}}
                ),
                Arguments.of(
                        "aquickbrownfoxjumpsoverthelazydog",
                        new String[]{"quick", "fox", "dog"},
                        new int[][]{{1, 5},{11, 13}, {30, 32}}
                )
        );
    }

    @ParameterizedTest
    @MethodSource("testCases")
    void indexPairs(String text, String[] words, int[][] expected) {
        int[][] actual = IndexPairs.of(text, words);
        assertArrayEquals(expected, actual);
    }
}

File: algorithms/src/test/kotlin/com/kotlinground/algorithms/trie/indexpairs/IndexPairsTest.kt
package com.kotlinground.algorithms.trie.indexpairs

import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource

class IndexPairsTest {
    companion object {
        @JvmStatic
        fun testCases() = listOf(
            Arguments.of(
                "thestoryofeducativeandme",
                listOf("story", "feduc", "educative"),
                listOf(
                    Pair(3, 7),
                    Pair(9, 13),
                    Pair(10, 18),
                )
            ),
            Arguments.of(
                "xyxyx",
                listOf("xyx", "xy"),
                listOf(
                    Pair(0, 1),
                    Pair(0, 2),
                    Pair(2, 3),
                    Pair(2, 4),
                )
            ),
            Arguments.of(
                "howareyou",
                listOf("how", "are", "you"),
                listOf(
                    Pair(0, 2),
                    Pair(3, 5),
                    Pair(6, 8),
                )
            ),
            Arguments.of(
                "weather",
                listOf("weather"),
                listOf(
                    Pair(0, 6),
                )
            ),
            Arguments.of(
                "aquickbrownfoxjumpsoverthelazydog",
                listOf("quick", "fox", "dog"),
                listOf(
                    Pair(1, 5),
                    Pair(11, 13),
                    Pair(30, 32),
                )
            ),
        )
    }

    @ParameterizedTest(name =  "{index} => indexPairs(\"{0}\", \"{1}\") should return {2}")
    @MethodSource("testCases")
    fun `index pairs of a string`(s: String, words: List<String>, expected: List<Pair<Int, Int>>) {
        val actual = indexPairs(s, words)
        assert(actual == expected)
    }
}
